File: src/runtime/watcher.ts
/**
 * FleetMind watcher — GitOps skills watcher.
 *
 * Polls a versioned skills repo for changes and pushes updates to agent workspaces.
 */

import fs from "node:fs";
import path from "node:path";
import type { Fleet, AgentConfig } from "../config/schema.js";
import { log } from "../utils/log.js";

export interface SkillUpdate {
  agentId: string;
  skillName: string;
  oldVersion: string | undefined;
  newVersion: string;
  pinned: boolean;
}

const CACHE_DIR = path.join(process.env.HOME ?? "/tmp", ".fleetmind", "skills-cache");

export class SkillsWatcher {
  constructor(private fleet: Fleet) {}

  /** Fetch versions.json from the skills repo. Returns { skillName: version }. */
  async fetchVersions(): Promise<Record<string, string>> {
    const repo = this.fleet.skills_repo;

    // Local mode — read directly
    if (repo.local) {
      const vFile = path.resolve(repo.local, "versions.json");
      if (fs.existsSync(vFile)) {
        return JSON.parse(fs.readFileSync(vFile, "utf-8"));
      }
      return {};
    }

    // Remote mode — clone or pull
    if (!repo.url) return {};

    try {
      const { simpleGit } = await import("simple-git");
      const git = simpleGit();

      if (fs.existsSync(CACHE_DIR)) {
        await simpleGit(CACHE_DIR).pull();
      } else {
        fs.mkdirSync(CACHE_DIR, { recursive: true });
        await git.clone(repo.url, CACHE_DIR, ["--branch", repo.branch ?? "main"]);
      }

      const vFile = path.join(CACHE_DIR, "versions.json");
      if (fs.existsSync(vFile)) {
        return JSON.parse(fs.readFileSync(vFile, "utf-8"));
      }
    } catch (err) {
      log.warn(`Error fetching skills repo: ${String(err)}`);
    }

    return {};
  }

  /** Get installed skill versions for an agent by reading package.json files. */
  getInstalledVersions(agentId: string): Record<string, string> {
    const workspace = path.join(
      this.fleet.agents.defaults.workspace_base,
      `workspace-${agentId}`
    );
    const skillsDir = path.join(workspace, "skills");
    if (!fs.existsSync(skillsDir)) return {};

    const versions: Record<string, string> = {};
    for (const entry of fs.readdirSync(skillsDir, { withFileTypes: true })) {
      if (!entry.isDirectory()) continue;
      const pkgFile = path.join(skillsDir, entry.name, "package.json");
      if (fs.existsSync(pkgFile)) {
        try {
          const pkg = JSON.parse(fs.readFileSync(pkgFile, "utf-8"));
          versions[entry.name] = pkg.version ?? "unknown";
        } catch {
          versions[entry.name] = "unknown";
        }
      }
    }
    return versions;
  }

  /** Compare installed vs available versions for an agent. Respects pins. */
  diff(agentId: string, available: Record<string, string>): SkillUpdate[] {
    const agent = this.fleet.getAgent(agentId);
    if (!agent) return [];

    const installed = this.getInstalledVersions(agentId);
    const updates: SkillUpdate[] = [];

    for (const skillRef of agent.skills) {
      const { name, version: pin } = skillRef;
      const current = installed[name];
      const latest = available[name];
      if (!latest) continue;

      if (pin) {
        // Pinned — flag but don't auto-update
        if (current !== pin) {
          updates.push({ agentId, skillName: name, oldVersion: current, newVersion: pin, pinned: true });
        }
      } else {
        // Unpinned — update to latest
        if (current !== latest) {
          updates.push({ agentId, skillName: name, oldVersion: current, newVersion: latest, pinned: false });
        }
      }
    }

    return updates;
  }

  /** Push a skill to an agent workspace. */
  pushSkill(agentId: string, skillName: string): boolean {
    const agent = this.fleet.getAgent(agentId);
    if (!agent) {
      log.error(`Agent ${agentId} not found`);
      return false;
    }

    const workspace = path.join(
      this.fleet.agents.defaults.workspace_base,
      `workspace-${agentId}`
    );
    const dest = path.join(workspace, "skills", skillName);
    const src = this.findSkillSource(skillName);

    if (!src) {
      log.error(`Skill ${skillName} not found`);
      return false;
    }

    fs.mkdirSync(path.dirname(dest), { recursive: true });
    if (fs.existsSync(dest)) fs.rmSync(dest, { recursive: true });
    fs.cpSync(src, dest, { recursive: true });

    log.ok(`Pushed ${skillName} to ${agentId}`);
    return true;
  }

  private findSkillSource(skillName: string): string | undefined {
    const candidates: string[] = [];
    if (this.fleet.skills_repo.local) {
      candidates.push(path.resolve(this.fleet.skills_repo.local, skillName));
    }
    candidates.push(path.resolve("./skills", skillName));
    if (fs.existsSync(CACHE_DIR)) {
      candidates.push(path.join(CACHE_DIR, skillName));
    }
    return candidates.find((c) => fs.existsSync(c));
  }

  private parsePollInterval(): number {
    const s = this.fleet.skills_repo.poll_interval ?? "60s";
    if (s.endsWith("h")) return parseInt(s) * 3600 * 1000;
    if (s.endsWith("m")) return parseInt(s) * 60 * 1000;
    return parseInt(s) * 1000;
  }

  /** Start the polling watch loop. Runs until SIGINT. */
  async watch(onUpdate?: (update: SkillUpdate) => void): Promise<void> {
    const interval = this.parsePollInterval();
    log.info(`FleetMind Watch — polling every ${this.fleet.skills_repo.poll_interval}`);
    log.info(`  Skills repo: ${this.fleet.skills_repo.url || this.fleet.skills_repo.local || "(none)"}`);
    log.info("  Press Ctrl+C to stop.\n");

    const tick = async () => {
      const available = await this.fetchVersions();
      const keys = Object.keys(available);
      if (keys.length > 0) {
        log.dim(`Available: ${keys.map((k) => `${k}@${available[k]}`).join(", ")}`);
      } else {
        log.dim("No versions.json found in skills repo.");
      }

      for (const agent of this.fleet.agents.list) {
        const updates = this.diff(agent.id, available);
        for (const u of updates) {
          if (u.pinned) {
            log.warn(`↔ ${agent.emoji} ${agent.name}: ${u.skillName} pinned@${u.newVersion} (installed: ${u.oldVersion ?? "none"}) — skipping`);
          } else {
            log.info(`↑ ${agent.emoji} ${agent.name}: ${u.skillName} ${u.oldVersion ?? "none"} → ${u.newVersion}`);
            this.pushSkill(agent.id, u.skillName);
            onUpdate?.(u);
          }
        }
      }
    };

    await tick();

    return new Promise((resolve) => {
      const timer = setInterval(async () => {
        try { await tick(); } catch (e) { log.error(String(e)); }
      }, interval);

      process.on("SIGINT", () => {
        clearInterval(timer);
        log.dim("\nWatcher stopped.");
        resolve();
      });
    });
  }
}

File: fleetmind/runtime/watcher.py
"""
FleetMind watcher — GitOps watcher for the versioned skills repo.

Polls a skills repo for version changes and pushes updated skills to agents.
"""

from __future__ import annotations
import json
import shutil
import time
from dataclasses import dataclass
from pathlib import Path
from typing import Callable

from rich.console import Console

from ..coordinator.config import FleetConfig, AgentConfig

console = Console()


@dataclass
class SkillUpdate:
    agent_id: str
    skill_name: str
    old_version: str | None
    new_version: str
    pinned: bool = False  # True = would update but is pinned


class SkillsWatcher:
    """Watches a versioned skills repo and pushes changes to agent workspaces."""

    def __init__(self, fleet: FleetConfig):
        self.fleet = fleet
        self._repo_cache_dir = Path("/tmp/fleetmind-skills-cache")

    def fetch_versions(self) -> dict[str, str]:
        """
        Fetch the versions.json index from the skills repo.
        Returns {skill_name: version} dict.

        Skills repo layout:
          versions.json  ← {\"coding\": \"1.2.0\", \"github\": \"2.1.0\", ...}
          coding/SKILL.md
          coding/package.json
          ...
        """
        repo = self.fleet.skills_repo

        # Local mode — read directly
        if repo.local:
            versions_file = Path(repo.local) / "versions.json"
            if versions_file.exists():
                return json.loads(versions_file.read_text())
            return {}

        # Remote mode — clone/pull the repo
        if not repo.url:
            return {}

        try:
            import git  # gitpython
            if self._repo_cache_dir.exists():
                r = git.Repo(self._repo_cache_dir)
                r.remotes.origin.pull()
            else:
                git.Repo.clone_from(repo.url, self._repo_cache_dir, branch=repo.branch)

            versions_file = self._repo_cache_dir / "versions.json"
            if versions_file.exists():
                return json.loads(versions_file.read_text())
        except ImportError:
            console.print("[yellow]gitpython not installed — cannot fetch remote skills repo[/yellow]")
        except Exception as e:
            console.print(f"[red]Error fetching skills repo:[/red] {e}")

        return {}

    def get_installed_versions(self, agent_id: str) -> dict[str, str]:
        """
        Read installed skill versions from agent workspace.
        Checks {workspace}/skills/{skill}/package.json for version field.
        Returns {skill_name: version}.
        """
        agent = self.fleet.get_agent(agent_id)
        if not agent:
            return {}

        workspace = Path(self.fleet.defaults.workspace_base) / f"workspace-{agent_id}"
        skills_dir = workspace / "skills"
        if not skills_dir.exists():
            return {}

        installed = {}
        for skill_dir in skills_dir.iterdir():
            if not skill_dir.is_dir():
                continue
            pkg_file = skill_dir / "package.json"
            if pkg_file.exists():
                try:
                    pkg = json.loads(pkg_file.read_text())
                    installed[skill_dir.name] = pkg.get("version", "unknown")
                except Exception:
                    installed[skill_dir.name] = "unknown"

        return installed

    def diff(self, agent_id: str, available: dict[str, str]) -> list[SkillUpdate]:
        """
        Compare installed skill versions against available versions.
        Returns list of SkillUpdate objects (respects version pins).
        """
        agent = self.fleet.get_agent(agent_id)
        if not agent:
            return []

        installed = self.get_installed_versions(agent_id)
        updates = []

        for skill_ref in agent.skills:
            name = skill_ref.name
            current = installed.get(name)
            latest = available.get(name)

            if not latest:
                continue  # Skill not in repo

            if skill_ref.version:
                # Pinned — flag but don't auto-update
                if current != skill_ref.version:
                    updates.append(SkillUpdate(
                        agent_id=agent_id,
                        skill_name=name,
                        old_version=current,
                        new_version=skill_ref.version,
                        pinned=True,
                    ))
            else:
                # Unpinned — update to latest
                if current != latest:
                    updates.append(SkillUpdate(
                        agent_id=agent_id,
                        skill_name=name,
                        old_version=current,
                        new_version=latest,
                        pinned=False,
                    ))

        return updates

    def push_skill(self, agent_id: str, skill_name: str, version: str | None = None) -> bool:
        """
        Push a skill to an agent workspace.
        Source: local skills dir or repo cache.
        Returns True if successful.
        """
        agent = self.fleet.get_agent(agent_id)
        if not agent:
            console.print(f"[red]Agent {agent_id} not found[/red]")
            return False

        workspace = Path(self.fleet.defaults.workspace_base) / f"workspace-{agent_id}"
        dest = workspace / "skills" / skill_name

        # Find source
        source = self._find_skill_source(skill_name)
        if not source:
            console.print(f"[red]Skill {skill_name} not found[/red]")
            return False

        dest.parent.mkdir(parents=True, exist_ok=True)
        if dest.exists():
            shutil.rmtree(dest)
        shutil.copytree(source, dest)

        console.print(f"[green]✓[/green] Pushed [bold]{skill_name}[/bold] to {agent_id}")
        return True

    def _find_skill_source(self, skill_name: str) -> Path | None:
        """Find skill source directory (local first, then repo cache)."""
        candidates = []

        if self.fleet.skills_repo.local:
            candidates.append(Path(self.fleet.skills_repo.local) / skill_name)

        candidates.append(Path("./skills") / skill_name)

        if self._repo_cache_dir.exists():
            candidates.append(self._repo_cache_dir / skill_name)

        for c in candidates:
            if c.exists():
                return c

        return None

    def _poll_interval_seconds(self) -> int:
        """Parse poll_interval string (e.g. '60s', '5m') to seconds."""
        interval = self.fleet.skills_repo.poll_interval
        if interval.endswith("s"):
            return int(interval[:-1])
        if interval.endswith("m"):
            return int(interval[:-1]) * 60
        if interval.endswith("h"):
            return int(interval[:-1]) * 3600
        return 60

    def watch(self, callback: Callable[[SkillUpdate], None] | None = None) -> None:
        """
        Polling loop. Fetches versions, diffs each agent, calls callback on changes.
        Runs until interrupted (Ctrl+C).
        """
        interval = self._poll_interval_seconds()
        console.print(f"[bold cyan]FleetMind Watch[/bold cyan] — polling every {interval}s")
        console.print(f"  Skills repo: {self.fleet.skills_repo.url or self.fleet.skills_repo.local or '(none)'}")
        console.print("  Press Ctrl+C to stop.\n")

        while True:
            try:
                available = self.fetch_versions()
                if available:
                    console.print(f"[dim]Available skills: {', '.join(f'{k}@{v}' for k,v in available.items())}[/dim]")
                else:
                    console.print("[dim]No skills repo version data found.[/dim]")

                for agent in self.fleet.agents:
                    updates = self.diff(agent.id, available)
                    for update in updates:
                        if update.pinned:
                            console.print(
                                f"[yellow]↔[/yellow] {agent.emoji} {agent.name}: "
                                f"{update.skill_name} pinned@{update.new_version} "
                                f"(installed: {update.old_version or 'none'}) — skipping auto-update"
                            )
                        else:
                            console.print(
                                f"[green]↑[/green] {agent.emoji} {agent.name}: "
                                f"{update.skill_name} {update.old_version or 'none'} → {update.new_version}"
                            )
                            self.push_skill(agent.id, update.skill_name, update.new_version)
                            if callback:
                                callback(update)

                time.sleep(interval)

            except KeyboardInterrupt:
                console.print("\n[dim]Watcher stopped.[/dim]")
                break
